src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -435,20 +440,49 @@ public async Task RemoveAccountAsync(RedirectContext context)
         /// </summary>
         internal /* for testing */ async Task<IConfidentialClientApplication> GetOrBuildConfidentialClientApplicationAsync()
         {
+            // Use a double checking lock to avoid locking in most situations
             if (_application == null)
             {
-                return await BuildConfidentialClientApplicationAsync().ConfigureAwait(false);
+                // Need to use an async lock because the lock keyword does not work in async functions 
+                await _applicationSyncObj.WaitAsync();
+                try
+                {
+                    if (_application == null)
+                    {
+                        _application = await BuildConfidentialClientApplicationAsync().ConfigureAwait(false);
+                    }
+                }
+                finally
+                {
+                    _applicationSyncObj.Release();
+                }
             }
 
             return _application;
         }
 
+        
+        private string BuildCurrentUriFromRequest(HttpContext httpContext, HttpRequest request)
+        {
+            // need to lock to avoid threading issues with code outside of this library
+            // https://docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?#do-not-access-httpcontext-from-multiple-threads
+            lock(httpContext)
```

Comment:
>httpContext [](start = 17, length = 11)

I'm not sure where values from the httpcontext is read here?

Reply:
Yes. Might be more readable to just lock on the `CurrentHttpContext` variable instead of passing around an HttpContext param.

Reply:
The example of what not to do in the docs link in the comment explicitly shows accessing a property on the HttpContext.Request object in parallel. The reason httpContext was passed in was mainly to guarantee that CurrentHttpContext was accessed only once, and to not have to do the null check again. I thought about locking on the httprequest object, but then realized that nearly all locks elsewhere will be done on the httpContext, so the highest chance of reducing threading issues would be to also lock on that.